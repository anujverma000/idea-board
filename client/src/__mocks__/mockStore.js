import configureStore from 'redux-mock-store'
import thunk from "redux-thunk";
import rootReducer from "../reducers";

const initialState= {
    idea: {
        list: []
    }
}
const mockStore = configureStore([thunk]);
const store = mockStore(initialState);
store.replaceReducer(rootReducer);

export default store;