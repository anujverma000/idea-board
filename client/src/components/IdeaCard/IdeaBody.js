import React, { useState } from "react";
import styled from "styled-components";

const IdeaBody = ({ idea, saveIdea }) => {

  const maxLength = 140;
  const displayLimit = 15;
  const [ ideaBody, setIdeaBody ] = useState(idea.body);
  const [ isEdited, setIsEdited ] = useState(false);
  const [ displayCharValue, setDisplayCharValue ] = useState(maxLength - idea.body.length);

  const updateIdea = () => {
    if (isEdited) {
      saveIdea(getUpdatedIdea());
      setIsEdited(false);
    }
  }

  const handleChange = e => {
    let value = e.target.value;
    setDisplayCharValue(maxLength - value.length);
    setIdeaBody(value);
    setIsEdited(true);
  }

  const getUpdatedIdea = () => ({...idea, body: ideaBody})

  return <>
          <IdeaBodyEditable
            value={ideaBody} 
            onChange={handleChange} 
            type="text"
            maxLength={maxLength}
            onBlur={updateIdea}
            placeholder="Description..."/>
          { displayCharValue < displayLimit && <CharCounter>{`${displayCharValue}/${maxLength}`}</CharCounter> }
        </>
}

/**
 *  STYLES
 */

const IdeaBodyEditable = styled.textarea`
  color: ${props => props.theme.textColor};
  font-size: 14px;
  font-weight: normal;
  resize: none;
  width: 95%;
  height: 85px;
  padding: 8px;
  border-radius: 4px;
  border: none;
  &:focus{
    outline: 1px solid ${props => props.theme.primaryColor};
  }
`;

const CharCounter = styled.div`
  right: 6px;
  bottom: 6px;
  position: absolute;
  font-size: 10px;
  background: #FFF1D8;
  padding: 4px 8px;
  border-radius: 16px;
  color: #562201;
  float: right;
  font-weight: bold;
`;

export default IdeaBody;