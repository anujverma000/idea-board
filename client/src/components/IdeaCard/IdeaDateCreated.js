import React from "react";
import styled from "styled-components";

const IdeaDateCreated = ({ date }) => {
    const getDisplayDateTime = timestamp => {
        let locale = 'en';
        let today = new Date();
        today.setHours(0);
        today.setMinutes(0);
        today.setSeconds(0, 0);

        let date = new Date(timestamp);
        date.setHours(0);
        date.setMinutes(0);
        date.setSeconds(0, 0);

        let differenceDays = (today - date)/86400000; /// 1000 * 3600 * 24

        let weeksFormat = { weekday: 'long'};
        let olderThanWeekFormat = { year: 'numeric', month: 'short', day: '2-digit' };
        let timeFormat = { hour:'2-digit', minute: '2-digit'};
        let displayDate = "";
        if (differenceDays === 0)
            displayDate = "Today";
        else if (differenceDays === 1)
            displayDate =  "Yesterday";
        else if(differenceDays > 1 && differenceDays < 7){
            displayDate =  date.toLocaleDateString(locale, weeksFormat);
        }else {
            displayDate =  date.toLocaleDateString(locale, olderThanWeekFormat);
        }
        return `${displayDate}, ${new Date(timestamp).toLocaleTimeString(locale, timeFormat)}`
    }
  return(
    <Wrapper>
      { getDisplayDateTime(date) }
    </Wrapper>
  )
}


/**
 *  STYLES
 */
const Wrapper = styled.div`
    font-size: 12px;
    margin-bottom: 12px;
`

export default IdeaDateCreated;