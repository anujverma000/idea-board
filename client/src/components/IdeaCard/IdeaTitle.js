import React, { useState } from "react";
import styled from "styled-components";

const IdeaTitle = ({ idea, saveIdea }) => {

  const maxLength = 140;

  const [ ideaTitle, setIdeaTitle ] = useState(idea.title);
  const [ isEdited, setIsEdited ] = useState(false);

  const saveIdeaUpdates = event => {
    if (event.keyCode === 13) {
      updateIdea();
    };
  }

  const updateIdea = () => {
    if (isEdited) {
      saveIdea(getUpdatedIdea());
      setIsEdited(false);
    }
  }

  const handleChange = e => {
    setIdeaTitle(e.target.value);
    setIsEdited(true);
  }

  const getUpdatedIdea = () => ({...idea, title: ideaTitle})

  return <TitleEditable
            value={ideaTitle} 
            onChange={handleChange} 
            type="text"
            maxLength={maxLength}
            autoFocus={ idea.new ? true: false }
            onKeyUp={saveIdeaUpdates}
            onBlur={updateIdea}
            placeholder="Add Idea"/>
}
/**
 *  STYLES
 */

const TitleEditable = styled.input`
  color: ${props => props.theme.textColorStrong};
  font-size: 16px;
  font-weight: bold;
  width: 95%;
  padding: 8px;
  border-radius: 4px;
  border: none;
  &:focus{
    outline: 1px solid ${props => props.theme.primaryColor};
  }
`
export default IdeaTitle;