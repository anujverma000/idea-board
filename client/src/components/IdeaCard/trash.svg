<svg width="24px" height="24px" viewBox="0 0 24 24" version="1.1">
<g id="surface1">
<path style=" stroke:none;fill-rule:nonzero;fill:rgb(7.45098%,27.058824%,38.823529%);fill-opacity:1;" d="M 9.1875 8.585938 L 10.3125 8.585938 L 10.3125 17.585938 L 9.1875 17.585938 Z M 9.1875 8.585938 "/>
<path style=" stroke:none;fill-rule:nonzero;fill:rgb(7.45098%,27.058824%,38.823529%);fill-opacity:1;" d="M 11.4375 8.585938 L 12.5625 8.585938 L 12.5625 17.585938 L 11.4375 17.585938 Z M 11.4375 8.585938 "/>
<path style=" stroke:none;fill-rule:nonzero;fill:rgb(7.45098%,27.058824%,38.823529%);fill-opacity:1;" d="M 13.6875 8.585938 L 14.8125 8.585938 L 14.8125 17.585938 L 13.6875 17.585938 Z M 13.6875 8.585938 "/>
<path style=" stroke:none;fill-rule:nonzero;fill:rgb(7.45098%,27.058824%,38.823529%);fill-opacity:1;" d="M 4.6875 5.210938 L 19.3125 5.210938 L 19.3125 6.335938 L 4.6875 6.335938 Z M 4.6875 5.210938 "/>
<path style=" stroke:none;fill-rule:nonzero;fill:rgb(7.45098%,27.058824%,38.823529%);fill-opacity:1;" d="M 14.773438 5.773438 L 13.726562 5.773438 L 13.726562 4.648438 C 13.726562 4.3125 13.460938 4.050781 13.125 4.050781 L 10.875 4.050781 C 10.539062 4.050781 10.273438 4.3125 10.273438 4.648438 L 10.273438 5.773438 L 9.226562 5.773438 L 9.226562 4.648438 C 9.226562 3.75 9.976562 3 10.875 3 L 13.125 3 C 14.023438 3 14.773438 3.75 14.773438 4.648438 L 14.773438 5.773438 "/>
<path style=" stroke:none;fill-rule:nonzero;fill:rgb(7.45098%,27.058824%,38.823529%);fill-opacity:1;" d="M 15.375 20.960938 L 8.625 20.960938 C 7.726562 20.960938 6.9375 20.210938 6.863281 19.3125 L 5.8125 5.8125 L 6.9375 5.738281 L 7.988281 19.238281 C 8.023438 19.574219 8.324219 19.835938 8.625 19.835938 L 15.375 19.835938 C 15.710938 19.835938 16.011719 19.539062 16.011719 19.238281 L 17.0625 5.738281 L 18.1875 5.8125 L 17.136719 19.3125 C 17.0625 20.25 16.273438 20.960938 15.375 20.960938 "/>
</g>
</svg>
