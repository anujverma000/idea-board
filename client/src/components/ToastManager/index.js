import React from "react";
const Ctx = React.createContext();

const ToastContainer = props => (
  <div style={{ position: "fixed", right: 0, bottom: 0 }} {...props} />
);

const Toast = ({ children, onDismiss }) => (
  <div
    style={{
      backgroundColor: "#16A085",
      color: "#FFFFFF",
      cursor: "pointer",
      fontSize: "16px",
      margin: "12px",
      padding: "12px 24px",
      borderRadius: "4px",
    }}
    onClick={onDismiss}
  >
    {children}
  </div>
);


export function ToastProvider({ children }) {
  const [toasts, setToasts] = React.useState([]);

  const showToast = content => {
    const id = Date.now();
    const toast = { content, id };
    setTimeout(() => removeToast(id), 3000);
    setToasts([toast]);
  };

  const removeToast = id => {
    const newToasts = toasts.filter(t => t.id !== id);
    setToasts(newToasts);
  };

  const onDismiss = id => () => removeToast(id);

  return (
    <Ctx.Provider value={{ showToast, removeToast }}>
      {children}
      <ToastContainer>
        {toasts.map(({ content, id, ...rest }) => (
          <Toast key={id} Toast={Toast} onDismiss={onDismiss(id)} {...rest}>
            {content}
          </Toast>
        ))}
      </ToastContainer>
    </Ctx.Provider>
  );
}


export const useToasts = () => React.useContext(Ctx);
