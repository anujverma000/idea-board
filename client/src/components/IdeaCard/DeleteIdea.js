import React from "react";
import { connect } from 'react-redux';
import styled from "styled-components";
import { ReactComponent as TrashIcon } from './trash.svg';
import { deleteIdea } from '../../actions/ideaActions';

const DeleteIdea = ({ idea, trashIdea }) => {

  const handleDelete = e => trashIdea(idea.id);
  
  return(
    <Wrapper onClick={handleDelete}>
      <TrashIcon/>
    </Wrapper>
  )
}


/**
 *  STYLES
 */
const Wrapper = styled.span`
  position: absolute;
  top: 8px;
  right: 8px;
`;

const mapDispatchToProps = dispatch => ({
  trashIdea: id =>  dispatch(deleteIdea(id))
});


export default connect(null, mapDispatchToProps)(DeleteIdea);