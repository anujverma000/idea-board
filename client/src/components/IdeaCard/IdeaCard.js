import React, { useState } from "react";
import { connect } from 'react-redux';
import styled from "styled-components";
import { useToasts } from "../ToastManager";
import IdeaBody from "./IdeaBody";
import IdeaTitle from './IdeaTitle'
import IdeaDateCreated from "./IdeaDateCreated";
import DeleteIdea from "./DeleteIdea";
import { updateIdea } from '../../actions/ideaActions';

const IdeaCard = ({ idea, updateIdea }) => {

  const [activeFocus, setActiveFocus] = useState(false);
  const [updatedIdea, setUpdatedIdea] = useState(idea);
  const { showToast } = useToasts();
  const saveIdea = ideaInstance => {
    setUpdatedIdea(ideaInstance);
    updateIdea(ideaInstance).then(() => {
      showToast("Saved Successfully.")
    });
  }

  return(
    <Wrapper onMouseEnter={ e => setActiveFocus(true) } onMouseLeave = { e => setActiveFocus(false) }>
      { activeFocus && <DeleteIdea idea={updatedIdea} /> } 
      <IdeaDateCreated date={ updatedIdea.created_date }/>
      <IdeaTitle idea={updatedIdea} saveIdea={saveIdea} />
      <IdeaBody idea={updatedIdea} saveIdea={saveIdea} />
    </Wrapper>
  )
}


/**
 *  STYLES
 */
const Wrapper = styled.div`
  padding: 16px;
  overflow-y: auto;
  background-color: ${props => props.theme.backgroundColor};
  border-radius: 4px;
  border: 1px solid ${props => props.theme.borderColor};
  box-sizing: border-box;
  height: 190px;
  overflow: hidden;
  position: relative;
  &:hover{
    box-shadow: 0 0 12px 4px ${props => props.theme.shadowColor};
    transition: .54s;
    transition-delay: .1s;
    cursor: pointer;
  }
`

const mapDispatchToProps = dispatch => ({
  updateIdea: ideaInstance => dispatch(updateIdea(ideaInstance))
});

export default connect(null, mapDispatchToProps)(IdeaCard);