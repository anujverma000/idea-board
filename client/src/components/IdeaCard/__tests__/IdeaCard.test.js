import React from 'react';
import renderer from 'react-test-renderer';
import IdeaCard from '../IdeaCard';
import mockStore from "../../../__mocks__/mockStore";
import { ToastProvider } from "../../ToastManager";
import { Provider } from "react-redux";

it('renders correctly', () => {

  const testIdea = {
    id: '123',
    title: 'Test Idea Title',
    body: 'Test Idea Body',
    created_date: 'Aug 2, 2019'
  };

  
  const tree = renderer.create(
      <Provider store={mockStore}>
        <ToastProvider>
          <IdeaCard idea={testIdea}/>
        </ToastProvider>
      </Provider>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
