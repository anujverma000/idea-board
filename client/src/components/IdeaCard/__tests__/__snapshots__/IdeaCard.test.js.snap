// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="sc-EHOje jesDKV"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="sc-bxivhb fLpZN"
    >
      Today, 12:00 AM
    </div>
    <input
      autoFocus={false}
      className="sc-htpNat inEueX"
      maxLength={140}
      onBlur={[Function]}
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="Add Idea"
      type="text"
      value="Test Idea Title"
    />
    <textarea
      className="sc-bdVaJa jTxIQE"
      maxLength={140}
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Description..."
      type="text"
      value="Test Idea Body"
    />
  </div>,
  <div
    style={
      Object {
        "bottom": 0,
        "position": "fixed",
        "right": 0,
      }
    }
  />,
]
`;
